rsky-pds/src/apis/com/atproto/repo/import_repo.rs

```
@@ -1,4 +1,157 @@
-#[rocket::post("/xrpc/com.atproto.repo.importRepo")]
-pub async fn import_repo() {
-    unimplemented!()
+use crate::actor_store::aws::s3::S3BlobStore;
+use crate::actor_store::ActorStore;
+use crate::apis::ApiError;
+use crate::auth_verifier::AccessFullImport;
+use crate::db::DbConn;
+use crate::repo::prepare::{
+    prepare_create, prepare_delete, prepare_update, PrepareCreateOpts, PrepareDeleteOpts,
+    PrepareUpdateOpts,
+};
+use aws_config::SdkConfig;
+use futures::{stream, StreamExt};
+use lexicon_cid::Cid;
+use rocket::data::ToByteUnit;
+use rocket::{Data, State};
+use rsky_common::env::env_int;
+use rsky_repo::block_map::BlockMap;
+use rsky_repo::car::read_stream_car_with_root;
+use rsky_repo::parse::get_and_parse_record;
+use rsky_repo::repo::Repo;
+use rsky_repo::sync::consumer::{verify_diff, VerifyRepoInput};
+use rsky_repo::types::{PreparedWrite, RecordWriteDescript, VerifiedDiff};
+
+#[tracing::instrument(skip_all)]
+#[rocket::post("/xrpc/com.atproto.repo.importRepo", data = "<blob>")]
+pub async fn import_repo(
+    auth: AccessFullImport,
+    blob: Data<'_>,
+    s3_config: &State<SdkConfig>,
+    db: DbConn,
+) -> Result<(), ApiError> {
+    let requester = auth.access.credentials.unwrap().did.unwrap();
+    let mut actor_store = ActorStore::new(
+        requester.clone(),
+        S3BlobStore::new(requester.clone(), s3_config),
+        db,
+    );
+
+    // Get current repo if it exists
+    let curr_root: Option<Cid> = actor_store.get_repo_root().await;
+    let curr_repo: Option<Repo> = match curr_root {
+        None => None,
+        Some(_root) => Some(Repo::load(actor_store.storage.clone(), curr_root).await?),
+    };
+
+    // Process imported car
+    let max_import_size = env_int("IMPORT_REPO_LIMIT").unwrap_or(100).megabytes();
```

Per [the documentation]:

>Import a repo in the form of a CAR file. Requires Content-Length HTTP header to be set.

It might take a little bit of work but one could fail early in the request if:
1. the content-length header is not set, 
2. the content-length header value is not a positve integer
3. the content-legnth header value exceeds `IMPORT_REPO_LIMIT` or 100 megabyes.  

and then it continues with the 
```rust
let requester = auth.access.credentials.unwrap().did.unwrap();
    let mut actor_store = ActorStore::new(
        requester.clone(),
        S3BlobStore::new(requester.clone(), s3_config),
        db,
    );
// code continues with rest of implementation ....
```

This would allow the server to ignore oversized requests right away without processing the entire stream of data.

---

I'll have to check, because from my own testing on my atproto-pds it didn't block me on content-length header. Do we want to utilize it to stay inline with the documentation, even if the atproto-pds doesn't follow suit?

---

Nice catch. I'll update it today accordingly as I agree

---

Updated and tested